libs/beacon/src/beacon_service.cpp

```diff
@@ -95,14 +99,56 @@ BeaconService::BeaconService(MuddleInterface &               muddle,
 
 BeaconService::Status BeaconService::GenerateEntropy(uint64_t block_number, BlockEntropy &entropy)
 {
-  FETCH_LOG_DEBUG(LOGGING_NAME, "Requesting entropy for block number: ", block_number);
+  FETCH_LOG_TRACE(LOGGING_NAME, "Requesting entropy for block number: ", block_number);
   beacon_entropy_last_requested_->set(block_number);
 
   FETCH_LOCK(mutex_);
 
-  if (completed_block_entropy_.find(block_number) != completed_block_entropy_.end())
+#if 1
+  {
+    static std::string last_have{};
+    std::ostringstream oss;
+
+    bool initial_loop{true};
+    for (auto const &element : completed_block_entropy_)
+    {
+      if (!initial_loop)
+      {
+        oss << ',';
+      }
+      oss << element.first;
+      initial_loop = false;
+    }
+
+    std::string now_have = oss.str();
+
+    if (now_have != last_have)
+    {
+      FETCH_LOG_INFO(LOGGING_NAME, "Have entropy for: ", oss.str());
+      last_have = std::move(now_have);
+    }
+  }
+#endif
+
+  auto it = completed_block_entropy_.find(block_number);
+  if (it != completed_block_entropy_.end())
   {
-    entropy = *completed_block_entropy_[block_number];
+    entropy = *(it->second);
+
+    // trim down the entropy cache
```

Comment:
I think this would cause the chain to stall indefinetly were there to be a significantly big fork. The completed entropy should be trimmed yes, however I would suggest switching to an ordered map and removing lower elements. Also signatures_being_built_ will 'leak' as well and should be cleared (I can do both of these things if you like)